### ISSUE TITLE
Refactor code and models to support TF 2.x and tf.keras

### ISSUE THREAD

**Issue by auroracramer** [marl/openl3#46]
At some point in the somewhat near future, we should establish support for TF 2.x and `tf.keras`. The main reasons for this are:
- To remain compatible with new releases of TF and Keras (the official version of which is now `tf.keras`) and make use of bug fixes, etc. some regression issues. As we have found (#42, #43), installing with newer versions of either package break installation and usage.
- To address multiple vulnerabilities contained in `tensorflow < 1.15.2`.
- To simplify the installation process; since TF 2.x now includes support for both CPU and GPU, we can now directly include tensorflow in the project dependencies, (as brought up in #39).

A priori, it seems like the main things to do are:
- Updating the dependencies in `setup.py` to include tensorflow
- Modifying the model definitions to be `tf.keras` compatible
- Porting the model files to a format that can be loaded by `tf.keras` with TF 2.x

The main concern that comes to mind is the regression tests. We have already seen that `tensorflow > 1.13` causes regression tests to fail. I imagine that this will only worsen as we introduce not only a new major release to TF, but also a divergence in Keras with `tf.keras`. @justinsalamon, what are your thoughts?

**Comment by justinsalamon** [comment]
Generally I totally support this move. My main concern as you've identified is the move from keras to `tf.keras`. As an example, I believe @jongwook tried to migrate CREPE from Keras to tf.keras and gave up on it because of incompatibilities and changes to the behavior of the model.

If we can make all of these changes while maintaining model performance (and hence regression tests), I think it's a no brainer. If the tests fail, then it'd probably be worth re-training the openl3 models against the new model implementation, ensuring the resulting embeddings are equally competitive on at least one downstream (e.g. US8K) and then releasing that.

Thoughts?

**Comment by justinsalamon** [comment]
p.s - we might be able to get away without re-training the models if we can at least show that the weights we have right now, coupled with the new implementation, still produce comparable downstream results. Then we'd just have to update the regression data for the regression tests.

So I guess my proposed steps would be:
1. Update to `tf.keras` - if regression tests pass, we're done.
2. If the regression tests fails, evaluate the resulting embeddings on downstream tasks. If they are equally competitive, update the regression test using the new implementation.
3. If the modified embeddings aren't competitive, then we'd have to re-train against the new implementation.

**Comment by jongwook** [comment]
> As an example, I believe @jongwook tried to migrate CREPE from Keras to tf.keras and gave up on it because of incompatibilities and changes to the behavior of the model.

This happened once when `tf.keras` was in the very early stage in the development, but [CREPE is currently using tf.keras](https://github.com/marl/crepe/blob/master/crepe/core.py#L44-L46) without causing disaster AFAIK. (and yes I am responsible for not writing real regression tests there)

**Comment by justinsalamon** [comment]
Ah cool, thanks for the info @jongwook. So it sounds like we should be able to at least give it a try, but yeah, we gotta be solid on performance and regression.

**Comment by jtcramer** [comment]
@justinsalamon That plan sounds good to me. I'll see about looking into this in the near future.

**Comment by thias42** [comment]
Hi, first of all thank you for sharing the code and model! I had a go at the TF2 support (branch tf2 in my fork) and it turns out that the regression tests fail after switching to tf.keras. So the new embeddings would need to be evaluated on any downstream tasks.

**Comment by gmos** [comment]
Hi, we are working with a sound recognizer currently based on vggish.  Our recognizer runs on an edge device and stretches the small CPU's to the limit. Too much heat and too much power usage.  So we want to try out tensorflow-lite, perhaps in combination with a Coral co-processor.  Unfortunately is only supported on TF2.  For vggish there are now some TF2 ports available and we have an evaluation of the port on our roadmap.

But we also very much would like to evaluate  openl3 for our application. But not if this would lock us to TF1.x again.
The last post for this issue was almost 4 months ago.  What are your current plans/goals regarding the TF2 port?

**Comment by justinsalamon** [comment]
@gmos we definitely plan to add TF 2.x support, but since this is a "side gig" for all of us, it is hard to provide concrete plans/estimates. We'll do our best to discuss in the near future and come up with an estimate, which we'll post here. Thanks!

**Comment by fluffynukeit** [comment]
@gmos When you say it's "only supported on TF2", do you mean vggish, tensorflow-lite, or the coral coprocessor?  I am considering using the coral coprocessor to run the openl3 model on TF 1.x.  Is that possible?

**Comment by gmos** [comment]
Nope. Last time when I looked at it, Coral required TF-Lite which only exists on TF2.

**Comment by fluffynukeit** [comment]
I was able to compile and run the openl3 model using tflite with tensorflow 1.x on google colab.  Here is the code to compile to tf-lite with %tensorflow_version 1.x.  I have to use colab because I don't have an Nvidia card and I could not get tflite to compile the openl3 model without it (after lots of attempts).  I can, however, run the openl3 models on my desktop machine, albeit very slowly.

```
import tensorflow as tf
import openl3

ctype = "env"
itype = "mel256"
esize = 512 # I'm only interested in 512 model, so don't change this

model = openl3.models.load_audio_embedding_model(
      content_type=ctype, embedding_size=esize, input_repr=itype)

# Note: getting all this working was tough to do.  tensorflow-gpu is essential
# and variables must be initialized.  Saving the model and/or loading the weights
# doesn't work.
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())

    bn = f"ol3_{ctype}_{itype}"
    converter = tf.compat.v1.lite.TFLiteConverter.from_session(
            sess, model.inputs, model.outputs)
    tflite_model = converter.convert()

    # Save the model.
    with open(f'{bn}.tflite', 'wb') as f:
      f.write(tflite_model)

```
I was able to download the tflite model file and then run it on my Desktop PC, which does not have CUDA or an Nvidia card, so it's running CPU only.  This code will run the tflite file.  I compared the output to the npz values generated with openl3 process_audio_file, and they don't agree at all, so there's either a bug or I am not preprocessing the data for tflite the same way process_audio_file does for keras.
```
import numpy as np
import tensorflow as tf
import soundfile as sf
import time

# Note: inputs data better be 48000 samples!
audio, sr = sf.read("1second.wav",
        dtype='float32') # dtype must match tflite model expectation

# Load the TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_path="ol3_env_mel256.tflite")
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Test the model on random input data.
input_shape = input_details[0]['shape']
print(input_shape)
audio = audio.reshape(1, 1, audio.shape[-1])

t = time.time()
input_data = audio
interpreter.set_tensor(input_details[0]['index'], input_data)
interpreter.invoke()

# The function `get_tensor()` returns a copy of the tensor data.
# Use `tensor()` in order to get a pointer to the tensor.
output_data = interpreter.get_tensor(output_details[0]['index'])
print(f"Time {time.time() - t} seconds")
print(output_data)

ol3_output = np.load("1second.npz")
ol3_embedding = ol3_output['embedding']
print(ol3_embedding)
```
I want to explore some more with getting TF lite to agree with the canned openl3 models as well as change the compile options for tf lite to use more integers.  It seems silly to require the audio input as float32 in my example when it's natively 16 bit signed audio.

**Comment by fluffynukeit** [comment]
I have edited my previous comment to include code that successfully compiles openl3 to TFLite in tensorflow 1.x on google colab.  TFlite model is faster and produces resulting audio embeddings that agree with vanilla openl3.

**Comment by justinsalamon** [comment]
Thanks @fluffynukeit !

We're hoping to port openl3 to TF 2.x, it's just a question of when we're able to allocate the cycles to it.

**Comment by fluffynukeit** [comment]
@justinsalamon  Thank you for providing openl3!  It saved me so, so much time.  I don't have proper ML hardware or ecosystem knowledge, so having models that largely "just work" out of the box for audio classification is great.  I threw random forest at the embeddings as suggested in another issue here and was able to iterate on tuning it instead of messing with the NN models.

**Comment by jonnor** [comment]
Since May this year and merge request #62 it looks like TF2 is supported. So when 0.4.0 is released maybe this issue can be closed?

**Comment by auroracramer** [comment]
Closed by #62
